---
title: "annotation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("/Users/jason/Documents/git_repos/Diversifying-the-pipeline-for-identifying-bulk-RNA-seq-derived-biomarkers-of-cancer-risk-within-sing/cluster_annotation")
```


Load in Seurat object containing integrated and normalized data from 8 donors
```{r loading data}
donors.integrated=readRDS("seurat-integrated.rds")
```


```{r defining clusters}
donors.integrated <- FindNeighbors(donors.integrated, dims = 1:10) # generates KNN graph
donors.integrated <- FindClusters(donors.integrated, resolution = 0.5) #louvain

#clusters can be accessed using Idents()
#cluster labels accessed via: levels(object)
```


```{r initial Cluster labeling on umap}
jpeg("seurat-cluster_number-umap.jpeg")
DimPlot(donors.integrated, reduction = "umap", group.by ="seurat_clusters" )
dev.off()
```

consider running this in parallel
```{r biomarkers for clusters }
#to run findallmarkers in parallel
#plan("multiprocess", workers = detectCores())

#find markers for all clusters
integrated.markers <- FindAllMarkers(donors.integrated, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)

#save
fwrite(integrated.markers, file="seurat-all_markers.txt")
#remove mitochondrial genes
integrated.markers=subset(integrated.markers,
                          !grepl("MT-",row.names(integrated.markers)))



topmarker=integrated.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)
```


```{r visualizing clusters with markers}
DimPlot(donors.integrated, reduction = "umap",group.by="donor")
DimPlot(donors.integrated, reduction = "umap",group.by="seurat_clusters")

#histogram for each cluster
RidgePlot(donors.integrated,features=topmarker$gene[6:10])

#dotplot: expression  + %age expressing 
DotPlot(donors.integrated,features=topmarker$gene) + RotatedAxis()

#Featureplot  #3d map with single gene heatmap


#heatmap with cluster abels and selected genes
DoHeatmap(donors.integrated, features = topmarker, size = 3)
```



find markers for all clusters. not all clusters may have genes that meet the threshold criteria (logFC & %age of cell expression)
```{r find positive markers for labeling}
plan("multiprocess", workers = detectCores())

markers <- FindAllMarkers(donors.integrated, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.58) #1.5fold

positive_markers=integrated.markers[integrated.markers$avg_logFC>0,]
#save
fwrite(positive_markers, file="seurat-positive_markers.txt")

#read data if needed
#positive_markers= read.table("/Users/jason/Documents/git_repos/Diversifying-the-pipeline-for-identifying-bulk-RNA-seq-derived-biomarkers-of-cancer-risk-within-sing/cluster_annotation/seurat-positive_markers.txt", sep = ",", header=TRUE)
```


## Cluster annotation using markers

files downloaded from CellMarker database
```{r loading cellmarker db markers}
human_markers=read.table("Human_cell_markers.txt", sep='\t',header = TRUE, stringsAsFactors = F)
single_markers=read.table("Single_cell_markers.txt", sep='\t',header = TRUE, stringsAsFactors = F)
#markers=read.table("cluster_markers.txt") #load saved marker data if needed
```


```{r checking for occurrence of db markers in biomarker genes}
#matching based on human cell markers
hmatch=merge(positive_markers,
             human_markers[,c("tissueType","cellName","geneSymbol")],by.x="gene",by.y="geneSymbol") 

#sort by cluster and padj
hmatch=arrange(hmatch,cluster, p_val_adj)

```

1,2,5, 7, 9, 10, 11 categories are part of cancer stem cell

0,3,4,6,8
```{r adding in info from single cell marker dataset}
sc_match=sapply(unique(positive_markers$gene),
       FUN = function(marker)
  single_markers[grep(marker,single_markers$cellMarker),"cellName"], simplify=T, USE.NAMES = T)

hmatch$sc_cellname=  apply( hmatch, 1, function(row) sc_match[row['gene']][1] )

#relevant columns and order
rel_col=c('cluster','gene','avg_logFC','pct.1','pct.2','cellName','sc_cellname')

View(hmatch[,rel_col])
#hdt=as.data.table(hmatch)

#View( hmatch[,by= cluster]   )


```


cluster labels:
0) FABP5; astrocyte
1)sox9, cd44 NES neural progenitor cell /cancer stem cell 
2) NES neural progenitor cell/ cancer stem cell
3) lgals3 , associated with cancer
4)
5) macrophages pecam1, cxcr4, cd68, cd14, eng(endoglin), icam1
7) cd24 neuroblast; (coexpression of nanog?)
9) cd24/dclk1
10) endothelial cells: cd34/pecam1/vwf/ mcam/ace/
11) thy1/eng mcam/eng




abcg2? cxcr4?


2) eosinophil
(cd9) schwann cell, astrocyte
3) B cell (cd74)
4) B cell ; cd74/gfap
5) macrophage (cd14/ cd68/cxcr4/cd163) with socx2/9 expression?
6) b cell cd74/gfap
7) cd24/ncam/cd9 ; neurons ncam=embryonic cells; cd24 (diff neuroblasts)
8) glul ; associated with liver carcinoma
9) sox2, nes, ncam1, thy1, cd24, abcg2 cancer; also has tissuetype: embryo
10) blood, thy1, cd9
11)


##


## Cluster annotation using Gene set analysis 
AUCell
https://bioconductor.org/packages/devel/bioc/vignettes/AUCell/inst/doc/AUCell.html#gene-sets


msigdb terms weren't very relevant
```{r download msigdb for gene sets}
#require(msigdbr)
#msigdbr_show_species()
m_df<- msigdbr(species = "Homo sapiens", category = "C7")


#divide table into list as defined by gene set
fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)

```


GENE ontology term loading
```{r download GO for gene sets and convert to symbols, message=FALSE, warning=FALSE}

require(pathview) #convert entrez gene to symbol
require(gage)

#gene sets are by default entrez id
go_gsets=go.gsets(species = "human", pkg.name=NULL, id.type = "eg", keep.evidence=FALSE)

#convert to symbols

require(future.apply)
plan("multiprocess", workers = detectCores())
#parallelize this
go_sym =future_lapply(go_gsets$go.sets, function(x){
  syms=eg2id(x, org="Hs", category="symbol")
  #syms will be a table with 2 columns ID & symbol, you just want the symbol
  return(syms[,2])
  
})
#go_sym will have 22,746 terms

remove(go_gsets)
```


```{r loading downsampled data aucell}
library(GSEABase)
library(AUCell)
require(Seurat)

#read in donors.integrated or down-sampled data
integrated.small=readRDS("/Users/jason/Documents/git_repos/Diversifying-the-pipeline-for-identifying-bulk-RNA-seq-derived-biomarkers-of-cancer-risk-within-sing/seurat-integrated-downsampled_100.rds")


exprMatrix=as.matrix(GetAssayData(integrated.small)) #extract expression matrix

umap_coord=Embeddings(integrated.small[["umap"]]) #get umap coordinates for downstream plotting

```

```{r remove gene sets with fewer than 20% genes needed for AUCell_calcAUC}

scgenes=rownames(exprMatrix)

go_satisf=future_lapply( go_sym, function(x){
if ( sum(x %in% scgenes) > 0.2*length(x)){
  return(x)
  } else {return(NULL)}
  })
go_satisf=Filter(Negate(is.null), go_satisf) #remove null vectors
```



```{r AUCell gene set analysis, message=FALSE, warning=FALSE}
#require(doMC)
cells_rankings <- AUCell_buildRankings(exprMatrix, nCores=6, plotStats=TRUE)

#change this to parallel
cells_AUC= AUCell_calcAUC(go_satisf, cells_rankings)
```

```{r AUCell assignment}
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=FALSE, assign=TRUE, nCores=5,verbose=FALSE) 
saveRDS(cells_assignment, "cells_assignment.rds")

#cells_assignment=readRDS("cells_assignment.rds")
```

```{r}
#extract assignments
cellsAssigned <- lapply(cells_assignment, function(x) x$assignment)
assignmentTable <- reshape2::melt(cellsAssigned, value.name="cell")

assignmentMat <- table(assignmentTable[,2], assignmentTable[,"cell"])
#number of cells assigned to each geneset
gs_total=rowSums(assignmentMat) %>% sort(decreasing = TRUE)

#find top10 genesets with greatest # of cells assigned



#condense cells in the same geneset
#rows= genesets, columns= cells, binary assignment to a geneset, you may have more than 1 geneset per cell (and sometimes 0)
assignmentMat <- table(assignmentTable[,2], assignmentTable[,"cell"])
miniAssigMat <- assignmentMat[1:50,sample(1:ncol(assignmentMat),10)] #100 random cells

require(NMF)
aheatmap(miniAssigMat, scale="none", color="black", legend=FALSE)
#cells (bottom), gene set(right)


```




```{r plot thresholded values}
interesting_cat=c(
"GO:0072606 interleukin-8 secretion",
"GO:0002369 T cell cytokine production",
"GO:0002477 antigen processing and presentation of exogenous peptide antigen via MHC class Ib",
"GO:2000977 regulation of forebrain neuron differentiation",
"GO:0001974 blood vessel remodeling",
"GO:0072091 regulation of stem cell proliferation"
)

selectedThresholds <- getThresholdSelected(cells_assignment) #gets AUC thresholds for each genes set (593)
selectedThresholds=selectedThresholds[interesting_cat] #only use thresholds of selected gene sets


par(mfrow=c(3,2)) # Splits the plot 
for(geneSetName in names(selectedThresholds))
{
  nBreaks <- 5 # Number of levels in the color palettes
  # Color palette for the cells that do not pass the threshold
  colorPal_Neg <- grDevices::colorRampPalette(c("black","blue", "skyblue"))(nBreaks)
  # Color palette for the cells that pass the threshold
  colorPal_Pos <- grDevices::colorRampPalette(c("pink", "magenta", "red"))(nBreaks)
  
  # Split cells according to their AUC value for the gene set
  passThreshold <- getAUC(cells_AUC)[geneSetName,] >  selectedThresholds[geneSetName]
  if(sum(passThreshold) >0 )
  {
    aucSplit <- split(getAUC(cells_AUC)[geneSetName,], passThreshold)
    
    # Assign cell color
    cellColor <- c(setNames(colorPal_Neg[cut(aucSplit[[1]], breaks=nBreaks)], names(aucSplit[[1]])), 
    setNames(colorPal_Pos[cut(aucSplit[[2]], breaks=nBreaks)], names(aucSplit[[2]])))
    
    # Plot
    plot(umap_coord, main=geneSetName,
    sub="Pink/red cells pass the threshold",
    col=cellColor[rownames(umap_coord)], pch=16) 
  }
}
```













You may need to map symbols to entrez id for goana
```{r mapping symbols to entrez for goana}
require(org.Hs.eg.db)
#keytypes(org.Hs.eg.db) #prints fields available for matching
#keys: enter list of genes in the format specified in "keytype"
#column: returned column field
entrez=mapIds(org.Hs.eg.db,keys= markers$gene, keytype="SYMBOL", column=("ENTREZID"))

#save symbols in new column
markers$entrez=entrez


# rank by number of cells assigned to each cluster
ordered_clust=markers %>% group_by(cluster) %>% count %>% arrange(desc(n))



```

generate a list containing enriched gene sets for each cluster
```{r goana on each cluster}
require(limma)
require(GO.db)

go_ls=list()

for (i in 1:nrow(ordered_clust)){
  #goana on each row (aka cell cluster)

#subset cluster markers for total markers table  
c_markers=markers[(markers$cluster==ordered_clust$cluster[i]) & (markers$p_val_adj<=.05), ]

go.out <- goana(unique(c_markers$entrez[!is.na(c_markers$entrez)]), species="Hs",  universe=NULL)

#focus on biological processes:BP
#u can also limit size of gene set with column: N
go.useful <- go.out[go.out$Ont=="BP" &go.out$P.DE<0.1,]

#order by pvalue
go.out <- go.out[order(go.out$P.DE),] 

#names the list with the cluster number
go_ls[[ordered_clust$cluster[i]]]=go.useful

print(paste("GO on cluster:", ordered_clust$cluster[i]))
}

```

```{r Viewing enriched gene sets}
View(head(go_ls[[10]],20))
saveRDS(go_ls,"enriched_gene_sets_by_cluster.rds")
```
cluster 1: immune response
cluster 4: also immune response
cluster 5: immune response
cluster 6: ?
cluster 9: ?
cluster 10: ?




```{r}
panglao=read.csv("panglao_output_top1gene.tsv",sep="\t")
```





